hs-abci-examples/simple-storage/test/SimpleStorage/Test/E2ESpec.hs

```diff
@@ -1,63 +1,121 @@
-      let txReq = RPC.RequestBroadcastTxCommit
-                    { RPC.requestBroadcastTxCommitTx = Base64.fromBytes . encode $ tx
-                    }
-      deliverResp <- fmap RPC.resultBroadcastTxCommitDeliverTx . runRPC $ RPC.broadcastTxCommit txReq
-      let deliverRespCode = deliverResp ^. Resp._deliverTxCode
-      deliverRespCode `shouldBe` 0
+      let txOpts = TxOpts
+            { txOptsGas = 0
+            , txOptsSigner = makeSignerFromUser user1
+            }
+          tx = SS.UpdateCountTx
+                 { SS.updateCountTxUsername = "charles"
+                 , SS.updateCountTxCount = 4
+                 }
+      resp <- assertTx . runTxClientM $ updateCount txOpts tx
+      ensureResponseCodes (0,0) resp
 
     it "can make sure the synchronous tx transaction worked and the count is now 4" $ do
-      let queryReq = QueryArgs
-            { queryArgsData = SS.CountKey
-            , queryArgsHeight = 0
-            , queryArgsProve = False
-            }
-      ClientResponse{clientResponseData = Just foundCount} <- runRPC $ getCount queryReq
+      resp <-  assertQuery . RPC.runTendermintM rpcConfig $
+        getCount defaultQueryArgs { queryArgsData = SS.CountKey }
+      let foundCount = queryResultData resp
       foundCount `shouldBe` SS.Count 4
 
 --------------------------------------------------------------------------------
+-- Query Client
+--------------------------------------------------------------------------------
+
+getCount
+  :: QueryArgs SS.CountKey
+  -> RPC.TendermintM (QueryClientResponse SS.Count)
 
-getCount :: QueryArgs SS.CountKey -> RPC.TendermintM (ClientResponse SS.Count)
-getCount =
-  let apiP = Proxy :: Proxy ("simple_storage" :> SS.Api)
-  in genClient (Proxy :: Proxy RPC.TendermintM) apiP def
+getAccount
+  :: QueryArgs Address
+  -> RPC.TendermintM (QueryClientResponse Auth.Account)
+
+getCount :<|> getAccount =
+  genClientQ (Proxy :: Proxy m) queryApiP def
+  where
+    queryApiP :: Proxy (QApi SimpleStorageModules)
+    queryApiP = Proxy
+
+
+--------------------------------------------------------------------------------
+-- Tx Client
+--------------------------------------------------------------------------------
+
+txClientConfig :: ClientConfig
+txClientConfig =
+  let getNonce addr = do
+        resp <- RPC.runTendermintM rpcConfig $ getAccount $
+            defaultQueryArgs { queryArgsData = addr }
+        case resp of
+          QueryError e ->
+            if appErrorCode e == 2
+              then pure 0
+              else error $ "Unknown nonce error: " <> show (appErrorMessage e)
+          QueryResponse QueryResult{queryResultData} ->
+            pure $ Auth.accountNonce queryResultData
+
+  in ClientConfig
+       { clientGetNonce = getNonce
+       , clientRPC = rpcConfig
+       }
+
+type TxClientM = ReaderT ClientConfig IO
+
+runTxClientM :: TxClientM a -> IO a
+runTxClientM m = runReaderT m txClientConfig
```

Comment:
Any reason why this doesn't live in the utils lib? It's also in the nameservice tests.

Reply:
because it requires the api for the auth module and currently I didn't implement a way to do that without generating an api for an entire application, not just a single module. I will go back later to do this, not this PR